### PR TITLE
Upgrade to Pinot 0.12.1

### PR DESCRIPTION
**Pull request by elonazoulay**
## Description

Upgrade pinot libraries to the latest version.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    private static final Map<String, TransformFunctionType> TRANSFORM_FUNCTION_TYPE_MAP;
+
+    static
+    {
+        Map<String, TransformFunctionType> builder = new HashMap<>();
+        for (TransformFunctionType transformFunctionType : TransformFunctionType.values()) {
+            for (String alias : transformFunctionType.getAliases()) {
+                TransformFunctionType previousValue = builder.put(canonicalize(alias), transformFunctionType);
+                checkState(previousValue == null || previousValue == transformFunctionType, "Duplicate key with different values for alias '%s', transform function type '%s' and previous value '%s'", canonicalize(alias), transformFunctionType, previousValue);
```
Comment (electrum) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java:41]: Did you mean to use `==` for comparing `TransformFunctionType`? Does it not implement equality and we need to use identity?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static org.apache.pinot.common.function.TransformFunctionType.SCALAR;
+import static org.apache.pinot.core.operator.transform.function.TransformFunctionFactory.canonicalize;
+
+public class PinotTransformFunctionTypeResolver
```
Comment (electrum) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java:29]: Make utility class final

**Review by electrum — APPROVED** [review]
The documentation says we support "Pinot 0.10.0 or higher". Is that still true after this change?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotExpressionRewriter.java [hunk 3]
```diff
@@ -156,6 +158,13 @@ private static <T> Optional<T> applyRule(RewriteRule<T> rule, T object, Context
         return Optional.empty();
     }
 
+    private static Literal toLiteral(String value)
+    {
+        Literal literal = new Literal();
+        literal.setStringValue(value);
+        return literal;
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotExpressionRewriter.java:165]: Why not `Literal.stringValue()` method?

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java [hunk 4]
```diff
@@ -308,13 +308,6 @@ private void createAndPopulateTooManyRowsTable(TestingKafka kafka, TestingPinotC
                     .set("updatedAt", initialUpdatedAt.plusMillis(i * 1000).toEpochMilli())
                     .build()));
         }
-        // For pinot 0.11.0+: rows with null time column values are ingested
-        // Only add a null row with a null time column for pinot < 0.11.0
-        if (!isLatestVersion()) {
```
Comment (ebyhr) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java:313]: Can we remove `isLatestVersion()` method now?

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java (hunk 1 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java:41]: It's an enum, is that ok?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotExpressionRewriter.java (hunk 3 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotExpressionRewriter.java:165]: It is overridden and surrounds non-null values with single quotes and null with the string `null`. Adding a comment to explain.

**Comment by elonazoulay** [comment]
> The documentation says we support "Pinot 0.10.0 or higher". Is that still true after this change?

Updated, thanks for catching that!

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotExpressionRewriter.java (hunk 3 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotExpressionRewriter.java:165]: Thanks! Updating.

Comment (ebyhr) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotExpressionRewriter.java:165]: I mean the new `toLiteral` method seem the same as `org.apache.pinot.common.request.Literal#stringValue`. 
```java
  public static Literal stringValue(java.lang.String value) {
    Literal x = new Literal();
    x.setStringValue(value);
    return x;
  }
```

https://github.com/apache/pinot/blob/d9c4315ca14997e6e0300d04f788e07723875159/pinot-common/src/main/java/org/apache/pinot/common/request/Literal.java#L204-L208

Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotExpressionRewriter.java]: Yep, just updated it, thanks for that!

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    // Extracted from org.apache.pinot.core.operator.transform.function.TransformFunctionFactory::get
+    public static Optional<TransformFunctionType> getTransformFunctionType(FunctionContext function)
+    {
+        requireNonNull(function, "function is null");
+        TransformFunctionType transformFunctionType = TRANSFORM_FUNCTION_TYPE_MAP.get(canonicalize(function.getFunctionName()));
+        if (transformFunctionType != null) {
+            return Optional.of(transformFunctionType);
+        }
+        if (FunctionRegistry.containsFunction(function.getFunctionName())) {
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java:55]: `TransformFunctionFactory::get` uses canonicalized function name here. Which is the expected behavior?

**Review by xiangfu0 — APPROVED** [review]
lgtm

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java (hunk 5 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java:55]: It calls canonicalize under the hood (in FunctionRegistry::containsFunction) but I will canonicalize here as well just to be safe.

Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTransformFunctionTypeResolver.java]: Updated.

**Pull request merged**